### PR TITLE
add video doorbell functionality

### PR DESCRIPTION
**Pull request by normen**
This PR optionally makes a camera-ffmpeg camera behave like a video doorbell and adds a switch that can be used to trigger a video doorbell event (notification).

### CODE REVIEW

**Comment by NorthernMan54** [comment]
@normen How is this different from the existing 'motion' trigger photo notification?

**Comment by Samfox2** [comment]
@normen would be nice if you had asked me before taking parts of my videodoorbell plugin and merge it here. I added the switch as an external http interface and the motion sensor aa a 2nd possibility to trigger the cam.

**Comment by normen** [comment]
@NorthernMan54 The behavior of the doorbell is different. The message is different and the doorbell shows a different display, as well as buttons for opening any doors that are in the same room.

@Samfox2 Most of the relevant parts here are my code, in "your" plugin as well - if you remember I contributed them to "your" plugin as well. However I decided that "your" plugin is way too unstable to contribute further on it, it crashes on every third update. Also "your" plugin is basically just a badly maintained fork of camera-ffmpeg with this code in it.

**Comment by Samfox2** [comment]
The plugin is a fork of homebridge-ffmpeg as mentioned on the first page. I did‘t merge it as I do not want to flood the issues page of this plugin even more with additional doorbell questions. I also did not want to add an additional dependency of the http request stuff. The plugin itself is running very stable from my side and from other that heard. Problems that occured in the past are mostly related to different hardware settings. Feel free to open an issue if you get an unstable behavior.

**Comment by normen** [comment]
Ok, so theres no issue with me posting this here, thanks.

**Comment by llemtt** [comment]
to me video doorbell makes no sense without two way audio!

and a louder "ringer" also because the homekit notification it's ok only to know that you have missed someone ringing at your door...

there are lots of cheaper "not homekit" video doorbell on the market that are not webcam waiting to be integrated using homebridge

I think it has to be a different plugin with is own development path although "inheriting" from homebridge-camera-ffmpeg 

cheers

**Comment by normen** [comment]
Two way audio works with camera-ffmpeg.

**Comment by normen** [comment]
> So this adds the doorbell notification. But there should be an option to show and hide of the doorbell switch.

The doorbell function has no use without a way to trigger it? Or do I get something wrong here?

**Comment by normen** [comment]
I can add the changes no problem @donavanbecker - I'm just unsure if it really makes sense.

**Comment by california444** [comment]
+1
please merge...thx
I use this plugin together with a doorbell to get a picture notification when the doorbell rings.
Right now I need the motion:true "workaround" in order to receive push notifications when the doorbell rings, but the notification says "motion detected"....

**Comment by normen** [comment]
> +1
> please merge...thx
> I use this plugin together with a doorbell to get a picture notification when the doorbell rings.
> Right now I need the motion:true "workaround" in order to receive push notifications when the doorbell rings, but the notification says "motion detected"....

So how do you trigger that? Do you automate the motion trigger switch? Or is there another way to trigger the bell without an additional switch? Then I'd get why I should make the switch optional.

**Comment by NorthernMan54** [comment]
@normen Can we also update the change log ( this is new ) with the changes

**Comment by normen** [comment]
There, should be in now

**Comment by donavanbecker** [comment]
> There, should be in now, also bumped the version to 1.0.1.
> 
> Edit: Or should that be 1.1.0? It's a feature after all.

let's do 1.1.0

File: index.js [hunk 1]
```diff
@@ -66,6 +70,29 @@ ffmpegPlatform.prototype.didFinishLaunching = function() {
         cameraAccessoryInfo.setCharacteristic(Characteristic.FirmwareRevision, cameraConfig.firmwareRevision);
       }
 
+      if(cameraConfig.doorbell) {
+        var doorbellService = new Service.Doorbell(cameraName+" Doorbell");
+        doorbellService.getCharacteristic(Characteristic.ProgrammableSwitchEvent)
+        .on('get', function(callback){
```
Comment (NorthernMan54) [inline index.js:76]: Is implementing the GET event required?  This is already covered with the timeout function

File: index.js [hunk 2]
```diff
@@ -66,6 +70,29 @@ ffmpegPlatform.prototype.didFinishLaunching = function() {
...
+        .on('set', function(state, callback){
+            if(state){
+                cameraAccessory.getService(Service.Doorbell).getCharacteristic(Characteristic.ProgrammableSwitchEvent).setValue(0);
+                setTimeout(function(){
+                    switchService.getCharacteristic(Characteristic.On).updateValue(false);
+                }, 1000);
+            }
+            callback(null, state);
+        })
+        .on('get', function(callback){
```
Comment (NorthernMan54) [inline index.js:91]: Is this required?   The updateValue as part of the timeout covers this

Comment (normen) [reply index.js:91]: Well at least to create the ProgrammableSwitchEvent Characteristic.. I guess just having an addCharacteristic() works, I'll check.

**Comment by normen** [comment]
Yeah, you were right, all unneded, removed them.

**Comment by NorthernMan54** [comment]
@normen I have a passion against .on('get' event handlers as is most cases they are not needed, and can be replaced with the use of updateValue instead.   .on('get' event handlers slow down Homebridge.

**Comment by normen** [comment]
D'oh, posted the answer as an edit above - anyway, should be as lean and unobtrusive as possible now 🙂

**Comment by donavanbecker** [comment]
Now getting a udid error.

```

Cannot add a Service with the same UUID '00000049-0000-1000-8000-0026BB765291' as another Service in this Accessory without also defining a unique 'subtype' property.[39m
[31m    at Accessory._this.addService (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:276:17)[39m
[31m    at PlatformAccessory.addService (/usr/local/lib/node_modules/homebridge/src/platformAccessory.ts:92:41)[39m
[31m    at /usr/local/lib/node_modules/homebridge-camera-ffmpeg/index.js:92:25[39m
[31m    at Array.forEach (<anonymous>)[39m
[31m    at ffmpegPlatform.didFinishLaunching (/usr/local/lib/node_modules/homebridge-camera-ffmpeg/index.js:44:13)[39m
[31m    at HomebridgeAPI.emit (events.js:310:20)[39m
[31m    at HomebridgeAPI.signalFinished (/usr/local/lib/node_modules/homebridge/src/api.ts:251:10)[39m
[31m    at Server.start (/usr/local/lib/node_modules/homebridge/src/server.ts:159:14)[39m
[31m    at cli (/usr/local/lib/node_modules/homebridge/src/cli.ts:80:10)[39m
[31m    at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)[39m
[37m[5/12/2020, 22:36:39] [39mGot SIGTERM, shutting down Homebridge...
```

**Comment by normen** [comment]
> Now getting a UUID error, when enabling doorbell.
> 
> ```
> 
> Cannot add a Service with the same UUID '00000049-0000-1000-8000-0026BB765291' as another Service in this Accessory without also defining a unique 'subtype' property.�[39m
> �[31m    at Accessory._this.addService (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:276:17)�[39m
> �[31m    at PlatformAccessory.addService (/usr/local/lib/node_modules/homebridge/src/platformAccessory.ts:92:41)�[39m
> �[31m    at /usr/local/lib/node_modules/homebridge-camera-ffmpeg/index.js:92:25�[39m
> �[31m    at Array.forEach (<anonymous>)�[39m
> �[31m    at ffmpegPlatform.didFinishLaunching (/usr/local/lib/node_modules/homebridge-camera-ffmpeg/index.js:44:13)�[39m
> �[31m    at HomebridgeAPI.emit (events.js:310:20)�[39m
> �[31m    at HomebridgeAPI.signalFinished (/usr/local/lib/node_modules/homebridge/src/api.ts:251:10)�[39m
> �[31m    at Server.start (/usr/local/lib/node_modules/homebridge/src/server.ts:159:14)�[39m
> �[31m    at cli (/usr/local/lib/node_modules/homebridge/src/cli.ts:80:10)�[39m
> �[31m    at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)�[39m
> �[37m[5/12/2020, 22:36:39] �[39mGot SIGTERM, shutting down Homebridge...
> ```

Sure thats because of my code? The motion option uses the same name as the camera, without adding a suffix like "Doorbell". My suspicion is that its the motion part.

**Comment by normen** [comment]
Yes, it only happened when the motion switch was enabled as well, I added a "Subtype" to both switches, that solves the issue.

**Comment by california444** [comment]
> So how do you trigger that? Do you automate the motion trigger switch? Or is there another way to trigger the bell without an additional switch? Then I'd get why I should make the switch optional. But the motion detection switch isn't optional either, the whole functionality is - just like the doorbell one in the current implementation.

Yes, I use the motion trigger switch. I would expect the same mechanism for the doorbell.

**Pull request merged**